Bearded.Utilities/Algorithms/CoffmanGraham.cs

```diff
@@ -0,0 +1,190 @@
+
+            private static ImmutableList<ImmutableHashSet<T>> createLayers(
+                // ReSharper disable once SuggestBaseTypeForParameter
+                IDirectedAcyclicGraph<T> graph, IList<T> ordering, int maxLayerSize)
+            {
+                var layersReversed = new List<ImmutableHashSet<T>.Builder>();
+                var elementToLayer = new Dictionary<T, int>();
+
+                for (var i = ordering.Count - 1; i >= 0; i--)
+                {
+                    // We fill in the layers back-to-front, always choosing the highest layer for the current element,
+                    // such that all its successors are in a higher layer, and the layer has less than W elements.
+                    // Since we are using a list where the right-most layer is 0, we are looking for the lowest k such
+                    // that (1) each successor is in a layer lower than k, and (2) layer k has less than W elements.
+
+                    // Requirement (1)
+                    var highestSuccessorLevel = graph.GetDirectSuccessorsOf(ordering[i])
+                        .Select(elmt => elementToLayer[elmt] + 1)
+                        .Aggregate(0, Math.Max);
+
+                    // Requirement (2)
+                    var candidateLayer = highestSuccessorLevel;
+                    while (
+                        candidateLayer < layersReversed.Count && layersReversed[candidateLayer].Count == maxLayerSize)
+                    {
+                        candidateLayer++;
+                    }
+
+                    // Expand the list as needed
+                    while (candidateLayer >= layersReversed.Count)
+                    {
+                        layersReversed.Add(ImmutableHashSet.CreateBuilder<T>());
+                    }
+
+                    // Add the element to the layer
+                    layersReversed[candidateLayer].Add(ordering[i]);
+                    elementToLayer.Add(ordering[i], candidateLayer);
+                }
+
+                return ImmutableList.CreateRange(layersReversed.Select(b => b.ToImmutable()).Reverse());
+            }
+        }
+
+        public static Instance<T> InstanceForGraph<T>(IDirectedAcyclicGraph<T> graph, int maxLayerSize)
+            where T : IEquatable<T>
+        {
+            return new Instance<T>(graph, maxLayerSize, isGraphReduced: false);
+        }
+
+        public static Instance<T> InstanceForAlreadyReducedGraph<T>(IDirectedAcyclicGraph<T> graph, int maxLayerSize)
+            where T : IEquatable<T>
+        {
+            return new Instance<T>(graph, maxLayerSize, isGraphReduced: true);
+        }
+
+        private struct DecreasingNumberSequence : IComparable<DecreasingNumberSequence>, IComparable
+        {
+            private readonly int[] numbers;
+
+            private DecreasingNumberSequence(int[] numbers)
+            {
+                this.numbers = numbers;
+            }
+
+            public int CompareTo(object obj)
```

Comment:
Is the `IComparable` used anywhere? Since this is entirely internal, I don't see the use case for the null and default cases below.

Reply:
`MinBy` requires the class to be `IComparable`. However, agreed that since this is an internal object, we don't need the special case handling.